[Geo] Integrate Lucene's LatLonShape (BKD Backed GeoShapes) as default `geo_shape` indexing approach

nknize opened the pull request:
This PR exposes lucene's `LatLonShape` field as the default type in `GeoShapeFieldMapper`. To use the new indexing approach, simply set `"type" : "geo_shape"` in the mappings without setting any of the `strategy`, `precision`, `tree_levels`, or `distance_error_pct` parameters. Note the following when using the new indexing approach:

* `geo_shape` query does not support querying by `MULTIPOINT`.
* `LINESTRING` and `MULTILINESTRING` queries do not yet support `WITHIN` relation. This will be added in the future as a `buffered_query`.
* `CONTAINS` relation is not yet supported.
* The `tree`, `precision`, `tree_levels`, `distance_error_pct`, and `points_only` parameters are deprecated.
* This PR will be backported to 6.6 (with deprecation logging included) and a separate parameter removal PR will be opened for 7.0.

All other `geo_shape` features are supported.

### TODO
- [x] Implement `POINT` queries as a `BoundingBox` query with `minLat == maxLat` && `minLon == maxLon`
- [x] Implement `GEOMETRYCOLLECTION` query as a `BooleanQuery`
- [x] Fix any remaining failing tests
- [x] Add deprecation logging for `tree`, `precision`, `tree_levels`, `distance_error_pct`, and `points_only` parameters in this PR (to be backported to 6.x)
- [x] Update docs

relates #32039

elasticmachine commented:
Pinging @elastic/es-search-aggs

nknize commented:
This is a wip so I expect some yaml tests (or two) to fail but it is close. I wanted to get it opened to get some eyes, discussion, and iterations on it.

iverase commented on server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -383,27 +393,71 @@ protected Query doToQuery(QueryShardContext context) {
+            if (queryShape instanceof double[][]) {
+                geometryType = GeoShapeType.MULTIPOINT;
+            } else if (queryShape instanceof double[] || queryShape instanceof GeoPoint) {
+                geometryType = GeoShapeType.POINT;
```
I think we can offer a point query by reusing the Box query. We just need to make minLat == maxLat and minLon == maxLon.

nknize replied:
:+1:  I'll run some tests with that approach.  I was also considering implementing `GeometryCollection` query as a `BooleanQuery`; though we'll likely need to run some benchmarks to make sure performance is acceptable.

nknize replied:
Implemented this for `POINT` queries, and it will work quite nicely. Unfortunately I found (and fixed) a small bug in `LatLonShapeBoundingBoxQuery`. So I'll either need to upgrade the lucene version to commit the implementation or hold off until another PR.
I do not want to implement this approach for `MULTIPOINT` queries, however. It would require rewriting to a `BooleanQuery` which would be quite slow, and abusive, for even just a few points.

nknize commented:
retest this please

nknize commented:
Still trying to figure out why these doc changes are failing. @nik9000 any thoughts?

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -517,20 +582,97 @@ public void parse(ParseContext context) throws IOException {
+    }
+
+    private void indexFields(ParseContext context, Field[] fields) {
+        ArrayList<IndexableField> flist = new ArrayList<>(Arrays.asList(fields));
```
why are you making a copy instead of just `List<IndexableField> flist = Arrays.asList(fields)`?

nknize replied:
Because `createFieldNamesField` adds elements to the list, and Lists returned by `Arrays.asList` cannot be structurally modified.

jpountz replied:
thanks I had missed this

jpountz commented on server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -383,27 +393,86 @@ protected Query doToQuery(QueryShardContext context) {
+            // for now just create a single bounding box query with min values == max values
+            double[] pt = queryShape instanceof GeoPoint
+                ? new double[] {((GeoPoint)queryShape).lon(), ((GeoPoint)queryShape).lat()}
+                : (double[])queryShape;
```
should we fail if the length of the `double[]` is not 2?

nknize replied:
:+1:

jpountz commented on server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -383,27 +393,86 @@ protected Query doToQuery(QueryShardContext context) {
+        for (Object shape : shapes) {
+            bqb.add(getVectorQuery(context, shape), BooleanClause.Occur.SHOULD);
+        }
+        return bqb.build();
```
let's wrap in a ConstantScoreQuery?

nknize replied:
Its wrapped in a `ConstantScoreQuery` on line 432.

nknize commented on server/src/main/java/org/elasticsearch/common/geo/SpatialStrategy.java:
```diff
@@ -50,7 +51,7 @@ public void writeTo(StreamOutput out) throws IOException {
     public static SpatialStrategy fromString(String strategyName) {
         for (SpatialStrategy strategy : values()) {
-            if (strategy.strategyName.equals(strategyName)) {
+            if (strategy.strategyName.equalsIgnoreCase(strategyName)) {
```
¯\\_(ツ)_/¯     :-)  Will fix.

nknize commented on server/src/main/java/org/elasticsearch/common/geo/ShapeRelation.java:
```diff
@@ -62,6 +63,17 @@ public static ShapeRelation getRelationByName(String name) {
+            case DISJOINT: return QueryRelation.DISJOINT;
+            case WITHIN: return QueryRelation.WITHIN;
+            default:
+                throw new IllegalArgumentException("ShapeRelation [" + CONTAINS + "] not supported");
```
:+1:

nknize commented:
> I'm wondering that things would be a bit easier to read eg. if we renamed the current field mapper to something like LegacyGeoShapeFieldMapper

:+1: I did it this way when I migrated `geo_point` to `LatLonPoint`. It works nicely and given the discussion from the other day re: versioning and release timing I think its the right way to go.

nknize commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -598,7 +609,8 @@ POST /example/doc
 Elasticsearch supports a `circle` type, which consists of a center
-point with a radius:
+point with a radius. Note that circles can only be indexed when using
+the `recursive` Prefix Tree strategy only.
```
:+1:

nknize commented:
OK... this took a little bit of work but I have made the following changes:

1. refactored `GeoShapeFieldMapper` into:
   -- `LegacyGeoShapeFieldMapper` - provides all legacy Prefix Tree parameters, and uses legacy PrefixTree indexing provided by lucene's `spatial-extras` module
   -- `GeoShapeFieldMapper` - provides new vector based indexing approach provided by lucene's new `LatLonShape` field type
   -- `BaseGeoShapeFieldMapper` - provides support for common parameters used between `GeoShapeFieldMapper` and `LegacyGeoShapeFieldMapper` (e.g., `coerce`, `ignore_malformed`, `ignore_z_value`, `orientation`)

2. Removed the new `vector` strategy
   -- Since `strategy` is only needed for prefix trees and the default indexing approach for newly created indexes is `LatLonShape` there is no need to have users explicitly specify a `strategy` the ideal mapping from here forward is:

```
"properties" : {
  "location" : {
    "type" : "geo_shape"
  }
}
```
   No other parameters are necessary

3. updated documentation to remove vector strategy and better describe the new indexing approach

nknize commented:
This commit is over +3500/-1500 changes. I'll look at splitting into two separate PRs:

1. Adding support to `ShapeBuilders` for building the Lucene Geometry needed by the new `LatLonShape` indexing approach
2. Refactoring `GeoShapeFieldMapper` and updating `GeoShapeQuery`

jpountz commented:
That would be great!

nknize commented:
ShapeBuilder PR is open at #35707. Once it is merged I'll open up the second PR for refactoring `GeoShapeFieldMapper`

nknize commented:
Merged #35707 into this branch so I don't think there's a need to open a separate PR. This one has been updated with all the ShapeBuilder changes and includes just the Mapper, Query Builder, and all test changes.

jpountz commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -100,16 +107,33 @@ and reject the whole document.
 |=======================================================================
 
+
+[[indexing-approach]]
```
we need a more qualified name since this links may be used from anywhere in the docs, eg. `geo-shape-indexing-approach`

jpountz commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -100,16 +107,33 @@ and reject the whole document.
+[[indexing-approach]]
+[float]
+==== Indexing approach
+GeoShape types are indexed by decomposing the shape into a triangular mesh and
```
maybe we could mention that indexing performance mostly depends on the number of edges of the input (multi)polygon

nknize replied:
:+1:

jpountz commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -146,21 +171,21 @@ are provided:
 
-|`recursive` |<<input-structure, All>> |`INTERSECTS`, `DISJOINT`, `WITHIN`, `CONTAINS` |Yes
+|`recursive`  |<<input-structure, All>> |`INTERSECTS`, `DISJOINT`, `WITHIN`, `CONTAINS` |Yes
 |`term` |<<point, Points>> |`INTERSECTS` |Yes
```
we need a new line for `vector`?

nknize replied:
No. I went ahead and removed `vector` altogether and made BKD indexing the default. No `strategy` needed. Its now been deprecated along with all other PrefixTree specific parameters. Just made more sense.

jpountz commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -100,16 +107,33 @@ and reject the whole document.
+==== Indexing approach
+GeoShape types are indexed by decomposing the shape into a triangular mesh and
+indexing each triangle as a 7 dimension point in a BKD tree. This provides
+near perfect spatial resolution (down to 1e-6 decimal degree precision) since all
```
Isn't it `1e-7`? (`360 / 2^32` ~= `8.4e-8` ~= `1e-7`)

nknize replied:
:+1:  That's right... I was wrong.

jpountz commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -185,22 +208,23 @@ PUT /example
-this into a tree_levels setting of 20.
+This mapping definition maps the location field to the geo_shape
+type using the default vector implementation. It provides
+approximately 1e-6 decimal degree precision.
```
is it 1e-7 here too?

nknize replied:
:+1:

jpountz commented on docs/reference/migration/migrate_7_0/mappings.asciidoc:
```diff
@@ -52,3 +52,19 @@ as a better alternative.
+
+The following type parameters are deprecated for the `geo_shape` field type: `tree`,
+`precision`, `tree_levels`, `distance_error_pct`, `points_only`, and `strategy`. They
+will be removed in a future version.
```
these changes should be in `docs/reference/migration/migrate_6_6.asciidoc` instead? (only exists on the 6.x branch)

nknize replied:
Yes I agree.... I can do that in the backport? Or should I open a whole new PR to handle that?

jpountz commented on server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java:
```diff
@@ -197,9 +197,6 @@ public Object buildLucene() {
 
-        if (shapes.size() == 1) {
-            return shapes.get(0);
-        }
```
why did you remove this?

nknize replied:
Because for Lucene geometry the shape could be a point defined as a `double[]` which fails downstream casting.

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java:
```diff
@@ -0,0 +1,342 @@
+    protected Explicit<Boolean> coerce;
+    protected Explicit<Boolean> ignoreMalformed;
+    protected Explicit<Boolean> ignoreZValue;
+    protected Explicit<Orientation> orientation;
```
Why do we hold the orientation here while it was only in the field type previously?

nknize replied:
Because orientation is used to build the geometry regardless of whether its S4J or Lucene. So it made more sense to have the parameter be part of the builder.

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -73,554 +49,125 @@
-        }
-        if (includeDefaults || coerce.explicit()) {
-            builder.field(Names.COERCE, coerce.value());
+            throw new IllegalArgumentException("invalid shape type found [" + luceneShape.getClass().getName() + "] while indexing shape");
```
getClass().getName() will not be meaningful eg. for an anonymous class, let's just rely on the toString representation of the class?

nknize replied:
:+1:

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -73,554 +49,125 @@
+    }
+
+    private void indexShape(ParseContext context, Object luceneShape) {
+        if (luceneShape instanceof GeoPoint) {
```
having a root class doesn't make much sense from a Lucene perspective since you're not expected to manipulate them in a generic way. Maybe we could have a wrapper on the Elasticsearch side to add type safety that would look something like this? Would it address your concern?

```java
public class GeoShape {

  // exactly one of them is non-null
  public final double[] point;
  public final Line line;
  public final Polygon polygon;
  // other types we want to support

  public GeoShape(double[] point) {
    this.point = point;
    this.line = null;
    this.polygon = null;
  }

  public GeoShape(Line line) {
    this.point = null;
    this.line = line;
    this.polygon = null;
  }

  public GeoShape(Polygon polygon) {
    this.point = null;
    this.line = null;
    this.polygon = polygon;
  }
}
```

imotov replied:
I think this will just replace this huge `if (shape instanceof ... )` with `if (shape != null)` switch.

jpountz replied:
Indeed. I was trying to see whether that would address your concern that there is no root geometry class.

nknize replied:
@imotov perhaps this can be resolved in a future PR by using the coming Geo API?

imotov replied:
:+1:

jpountz commented on server/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java:
```diff
@@ -285,7 +284,8 @@ public void testParseMixedDimensionPolyWithHole() throws IOException {
         Mapper.BuilderContext mockBuilderContext = new Mapper.BuilderContext(indexSettings, new ContentPath());
-        final GeoShapeFieldMapper mapperBuilder = new GeoShapeFieldMapper.Builder("test").ignoreZValue(false).build(mockBuilderContext);
+        final LegacyGeoShapeFieldMapper mapperBuilder =
+            (LegacyGeoShapeFieldMapper)(new LegacyGeoShapeFieldMapper.Builder("test").ignoreZValue(false).build(mockBuilderContext));
```
should we test with the non-legacy mapper instead?

nknize replied:
:+1:

jpountz commented on server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java:
```diff
@@ -63,6 +63,7 @@
         private BooleanFieldMapper.Builder boolBuilder = new BooleanFieldMapper.Builder(Names.FIELD_BOOL);
         private GeoPointFieldMapper.Builder latLonPointBuilder = new GeoPointFieldMapper.Builder(Names.FIELD_POINT);
         private GeoShapeFieldMapper.Builder shapeBuilder = new GeoShapeFieldMapper.Builder(Names.FIELD_SHAPE);
+        private LegacyGeoShapeFieldMapper.Builder legacyShapeBuilder = new LegacyGeoShapeFieldMapper.Builder(Names.FIELD_SHAPE);
```
let's use the non-legacy mapper here too

nknize replied:
I have this there for backcompat testing. The `build()` method checks index version and uses this for indexes created before 6.6

jpountz replied:
ok thanks for explaining

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java:
```diff
@@ -0,0 +1,342 @@
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (DeprecatedParameters.parse(name, fieldName, fieldNode, deprecatedParameters)) {
```
do we want to disallow them on 7.0+ indices? I'm happy with it being done in a follow-up PR, just asking.

nknize replied:
:+1:  Yes. I was going to do this in a separate PR to simplify backporting these changes to 6.6

nknize commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -21,48 +21,55 @@ type.
+parameter defines the approach for how to represent shapes at indexing
+and search time. It also influences the capabilities available so it
+is recommended to let Elasticsearch set this parameter automatically.
+There are three strategies available: `recursive`, and `term`.
```
No. I need to change "three" to "two".  I removed the vector option altogether and just set `LatLonShape` as the default indexing strategy. This way users have to explicitly specify `tree` or `strategy` parameter to use the legacy (soon to be removed) PrefixTree approach. I thought this was cleaner.

nknize commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -21,48 +21,55 @@ type.
-may use the `precision` parameter instead. However, Elasticsearch only
-uses the tree_levels parameter internally and this is what is returned
-via the mapping API even if you use the precision parameter.
+|`tree_levels` |deprecated[6.6, PrefixTrees no longer used] Maximum number
```
:+1:

nknize commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -21,48 +21,55 @@ type.
-number followed by an optional distance unit. Valid distance units
-include: `in`, `inch`, `yd`, `yard`, `mi`, `miles`, `km`, `kilometers`,
-`m`,`meters`, `cm`,`centimeters`, `mm`, `millimeters`.
+|`precision` |deprecated[6.6, PrefixTrees no longer used] This parameter may
```
:+1:

nknize commented on docs/reference/mapping/types/geo-shape.asciidoc:
```diff
@@ -21,48 +21,55 @@ type.
-GeohashPrefixTree and `quadtree` for QuadPrefixTree.
+|`tree |deprecated[6.6, PrefixTrees no longer used] Name of the PrefixTree
+implementation to be used: `geohash` for GeohashPrefixTree and `quadtree`
+for QuadPrefixTree.
```
:+1:

nknize commented on server/src/test/java/org/elasticsearch/common/geo/GeoJsonShapeParserTests.java:
```diff
@@ -1078,6 +1077,7 @@ public void testParseGeometryCollection() throws IOException {
                         .endObject()
                     .endArray()
                 .endObject();
+        boolean useJTS = randomBoolean();
```
:+1:

nknize commented:
I think this is ready for another round of review. Just note that `vector` is no longer a strategy. It made more sense to make BKD the new default indexing option and PrefixTrees an "opt in" strategy. Also, `strategy` is a deprecated option since it's only relevant for Prefix Tree indexing. It didn't make much sense to add a new strategy to a soon to be removed parameter.

nknize commented:
retest this please

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapper.java:
```diff
@@ -0,0 +1,596 @@
+                DistanceUnit.METERS.toString(50));
+        }
+        if (includeDefaults || fieldType().strategy() != DeprecatedParameters.Defaults.STRATEGY) {
+            builder.field(DeprecatedParameters.Names.STRATEGY.getPreferredName(), fieldType().strategy().getStrategyName());
```
let's always have the strategy in the xcontent for legacy shapes to avoid any confusion? ie. remove the if statement and always execute the body

nknize replied:
:+1:  great idea

jpountz commented on server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -376,34 +382,99 @@ protected Query doToQuery(QueryShardContext context) {
+        if (relation == ShapeRelation.CONTAINS) {
+            throw new QueryShardException(context,
+                ShapeRelation.CONTAINS + " query relation not supported for Field [" + fieldName
+                    + "] Use prefix tree strategy [" + SpatialStrategy.RECURSIVE + "] or [" + SpatialStrategy.TERM + "] instead.");
```
it's a bit weird to recommend deprecated functionality?

nknize replied:
lol.. that's a great point. I'll remove the recommendation

nknize commented on server/src/main/java/org/elasticsearch/common/geo/parsers/GeoJsonParser.java:
```diff
@@ -102,7 +108,7 @@ protected static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper s
                         }
                         subParser.nextToken();
-                        requestedOrientation = ShapeBuilder.Orientation.fromString(subParser.text());
+                        orientation = ShapeBuilder.Orientation.fromString(parser.text());
```
:+1:

nknize commented:
Will keep an eye on the tests... but it looks like this is good to go.

philvarner commented:
One side effect of the change `Implement GEOMETRYCOLLECTION query as a BooleanQuery` seems to be that large geometry collections (my case, the System for Terrestrial Ecosystem Parameterization (STEP) geojson) that, prior to 6.6, were counted as a single clause, now seem to have each geometry count as a clause, and will hit the maxClauseCount default limit of 1024.  This was unexpected and something I couldn't find documented anywhere, so I thought it would be useful to at least comment here about it.
